azkaban-common/src/main/java/azkaban/scheduler/MissedSchedulesManager.java

```diff
@@ -0,0 +1,199 @@
+import azkaban.utils.Emailer;
+import azkaban.utils.Props;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+import java.text.MessageFormat;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Miss Schedule Manager is a failure recover manager for schedules. Schedule might be missed to execute due to
+ * Trigger Scanner thread busy or web server down. This manager maintains a configurable fixed sized thread pool
+ * to execute tasks including send email notification or back execute flow.
+ * */
+@Singleton
+public class MissedSchedulesManager {
+  private static final Logger LOG = LoggerFactory.getLogger(MissedSchedulesManager.class);
+
+  private ExecutorService missedSchedulesExecutor;
+  private final boolean missedSchedulesManagerEnabled;
+  private final ProjectManager projectManager;
+  // TODO: Refactored to AlerterHolder to support missed scheduler use case
+  private final Emailer emailer;
+
+  // parameters to adjust MissScheduler
+  private final int threadPoolSize;
+  private final int DEFAULT_THREAD_POOL_SIZE = 5;
+
+  @Inject
+  public MissedSchedulesManager(final Props azkProps,
+      final ProjectManager projectManager,
+      final Emailer emailer) {
+    this.projectManager = projectManager;
+    this.emailer = emailer;
+    this.threadPoolSize = azkProps.getInt(Constants.MISSED_SCHEDULE_THREAD_POOL_SIZE, DEFAULT_THREAD_POOL_SIZE);
+    this.missedSchedulesManagerEnabled = azkProps.getBoolean(Constants.MISSED_SCHEDULE_MANAGER_ENABLED, false);
+    if (this.missedSchedulesManagerEnabled && threadPoolSize <= 0) {
+      String errorMsg = "MissedSchedulesManager is enabled but thread pool size is not positive: " + threadPoolSize;
+      LOG.error(errorMsg);
+      throw new RuntimeException(errorMsg);
+    }
+  }
+
+  /**
+   * Put timestamps for Missed executions along with email notifications into Task Queue.
+   *
+   * @param missedScheduleTimesInMs, timestamps of missed schedule
+   * @param action, execute flow action
+   * @param backExecutionEnabled, a schedule config from user, the default is false
+   * */
+  public boolean addMissedSchedule(final List<Long> missedScheduleTimesInMs,
+      final ExecuteFlowAction action,
+      final boolean backExecutionEnabled) {
+    String projectId = action.getProjectName();
+    String flowName = action.getFlowName();
+    Project project = projectManager.getProject(projectId);
+    if (project == null) {
+      return false;
+    }
+    List<String> emailRecipients = project.getFlow(flowName).getFailureEmails();
```

Comment:
With `getFlow().getFailureEmails()` we'll probably only get failure emails defined in the yaml file but we'll miss any overwrites or new definitions added when the schedule was created. We need to check all possible sources.

Reply:
What do you mean that? If they change the flow failure emails it can not be overwrites?

Reply:
Failure emails can be defined in the .yaml/.properties files but also at the time of creating the schedule where runtime properties can be provided. I believe here we're only taking into consideration failure emails defined in the yaml but not possible overwrites or additions done through runtime properties and in fact those have higher precedence than the ones in the .yaml/.properties files.

Reply:
That's actually good call!! thank you, I didn't aware when defining schedules we provide them a failureEmail override. Adjust it in the code.